### PR TITLE
Support for manual viewport size input & reintroduce resize key combinations

### PR DESCRIPTION
**Pull request by JosefBredereck**
### Closes #

| Issue | Title |
|---|---|
| #1145 | Viewport resizing |

#### Features
- Ability to change viewport width by changing the value manually

### CODE REVIEW

**Comment by sghoweri** [comment]
Hey thanks so much for your help getting this original PL feature re-enabled @JosefBredereck! 

Starting to take a look at this...  👀

File: packages/uikit-workshop/src/scripts/lit-components/pl-viewport-size-list/pl-viewport-size-list.js [hunk 1]
```diff
@@ -74,20 +74,27 @@ class ViewportSizes extends BaseComponent {
...
-              maxViewportWidth
+              max
             ),
             true
           );
           break;
         case 'full':
           this.iframe.fullMode = true;
-          this.iframe.sizeiframe(maxViewportWidth, true);
+          this.iframe.sizeiframe(this.iframe.clientWidth, true);
```
Comment (sghoweri) [inline packages/uikit-workshop/src/scripts/lit-components/pl-viewport-size-list/pl-viewport-size-list.js:97]: I'm a bit iffy on the changes in this file... could you elaborate a bit more on the change in behavior here?

Let's say the user sets their max viewport width config to `1200px` but is viewing PL on a `2560px` wide display... if they hit the `full` button, shouldn't Pattern Lab's viewport resizer width max out at `1200px` vs resizing to `2560px`?

@bmuenzenmeyer @bradfrost thoughts here?

Comment (JosefBredereck) [reply packages/uikit-workshop/src/scripts/lit-components/pl-viewport-size-list/pl-viewport-size-list.js:97]: Shure, I can add some insights about my thoughts. I always felt that the `full` button is something like a reset to a state my screen supports. And that also means if I set large to a maximum of 1200px the full button will still scale to 2560px when I click on it. Also, there are cases where I need to test this behavior, especially when I want to test if my containers work correctly and even if the screen has 4k resolution, they do not size over a specific point.

**Comment by JosefBredereck** [comment]
That's some valid points. 

> Unlike the original implementation of this functionality, you can no longer press the up or down arrow keys. Ideally the behavior I think we’d want to see here would be: 

> Entering in a valid number into the input should automatically trigger the viewport to resize (or at the very least, entering a number into the input and tabbing away) — ie. I feel like users should be able to change the viewport size by simply changing the number (vs requiring them to hit enter)

Yes and No. I also tested that behavior. But let's imagine you are at `500px` and want to change it to `1000px`. When you start to replace the number you start at `1px` and the screen will be resized to `240px` minimum. After that you go to `10px` and `100px` and nothing changes. Then you reach `1000px` and the screen will jump to `1000px`. You can come across with a `debounce` function - BUT - that will conflict again with changing the numbers by clicking up and down. Will the number be evaluated when the user is done with going up and down or do we have a different behavior for this functionality?

> Deleting the text from the current viewport size and tabbing away should reset the displayed value to reflect the current viewport resized width

Should be fixed before merge.

> If the user inputs a value that’s outside of the min/max range of viewport sizes (and hits enter since that’s the current behavior on this PR branch I’m reviewing) — the displayed viewport size should correct itself to that min/max size vs displaying a value that’s out of bounds… Or to put it another way, the viewport resizer width should always be 100% in sync with the displayed size!

Therefore we need Blur the input field when pressing enter so that the input value will start the sync again.

> Side note: FWIW, I’ve never been a fan of the em option that displays next to the viewport size in pixels. That said, are users going to be confused by being able to (once again) set one input but not the other?

Actually that's something I was always wondering about. Or at least why that em is even there.

File: packages/uikit-workshop/src/scripts/lit-components/pl-viewport-size/pl-viewport-size.js [hunk 2]
```diff
@@ -52,12 +60,29 @@ class ViewportSize extends BaseLitComponent {
...
     if (!window.__PRERENDER_INJECTED) {
       return html`
-        <form class="pl-c-viewport-size" method="post" action="#">
-          <span class="pl-c-viewport-size__input">${this.px}px</span
-          >&nbsp;/&nbsp;
+        <form class="pl-c-viewport-size" @submit=${this.handleSubmit}>
+          <input
+            class="pl-c-viewport-size__input pl-c-viewport-size__input-action"
+            .value="${this.state.inputPixelValue}"
+            @keypress=${this.handlePixelUpdate}
```
Comment (JosefBredereck) [inline packages/uikit-workshop/src/scripts/lit-components/pl-viewport-size/pl-viewport-size.js:84]: Also, I need to mention that I'm not that well known with React. Maybe there is a better approach for handling an input and replace/prevent an invalid key. In Angular, there are directives for that.

File: packages/uikit-workshop/src/scripts/lit-components/pl-viewport-size-list/pl-viewport-size-list.js (hunk 1 above)
Comment (JosefBredereck) [reply packages/uikit-workshop/src/scripts/lit-components/pl-viewport-size-list/pl-viewport-size-list.js:97]: Little addition: I just found the file `view-sizes.js` there is a check if the value `window.config.ishMaximum` exists. I never saw this value in any `pl-config` (just available in `cli/test/util/patternlab-config.json`) file that's why I wasn't aware of it. Now it makes more sense and this changed line is invalid.

**Comment by JosefBredereck** [comment]
@sghoweri I added some more functionality. Please have a look at these too. 👍

File: packages/uikit-workshop/src/scripts/lit-components/pl-viewport-size/pl-viewport-size.js [hunk 3]
```diff
@@ -52,13 +62,40 @@ class ViewportSize extends BaseLitComponent {
...
+            <input
+              class="pl-c-viewport-size__input pl-c-viewport-size__input-action"
+              id="pl-viewport-width"
+              .value="${this.state.inputPixelValue}"
+              @keypress=${this.handlePixelUpdate}
+              @blur=${this.handlePixelBlur}
+            />px&nbsp;/&nbsp;<span class="pl-c-viewport-size__input"
+              >${this.em}em</span
+            >
+          </label>
```
Comment (JosefBredereck) [inline packages/uikit-workshop/src/scripts/lit-components/pl-viewport-size/pl-viewport-size.js:98]: I hope that's what you meant by wrapping in in a label.

Comment (sghoweri) [reply packages/uikit-workshop/src/scripts/lit-components/pl-viewport-size/pl-viewport-size.js:98]: 👍 yessir, exactly what I meant

File: packages/uikit-workshop/src/scripts/lit-components/pl-viewport-size-list/pl-viewport-size-list.js (hunk 1 above)
Comment (bmuenzenmeyer) [reply packages/uikit-workshop/src/scripts/lit-components/pl-viewport-size-list/pl-viewport-size-list.js]: I think I can help here.

The [Ish](https://bradfrost.com/blog/post/ish/) feature is not about setting a maximum that may never be exceeded. We all know how futile it is to imagine a world where we control the viewport of our users.

quoting from https://patternlab.io/docs/advanced-config-options.html

#### ishViewportRange
Sets the boundaries of each of the viewport toggles, `S`mall, `M`edium, and `L`arge. Clicking on one of these buttons will randomly set the ish Viewport to a value within the given range. Setting the range to the same number can effectively set an exact value. The first entry in `ishViewportRange.s` is the `ishViewportMinimum`, which is now obsolete. The second entry in `ishViewportRange.l` is the `ishViewportMaximum`, which is now also obsolete.

-----

to me, full has always meant full-screen, regardless of ish setting.

**Comment by bmuenzenmeyer** [comment]
> > Side note: FWIW, I’ve never been a fan of the em option that displays next to the viewport size in pixels. That said, are users going to be confused by being able to (once again) set one input but not the other?
> 
> Actually that's something I was always wondering about. Or at least why that em is even there.

One of the primary use cases of the viewport resizer is to exactly test media queries.

I can almost certainly guarantee `em` is there because of this: https://css-tricks.com/em-based-media-queries-are-based-on/

Chris makes it sound as if this is no longer needed, which TIL.

**Comment by JosefBredereck** [comment]
Oh no, I merged the wrong dev branch here 😢

**Comment by JosefBredereck** [comment]
It can not be saved, I close this one and make a new 🤦

**Pull request closed without merging**